File: sim/VolleySim.Core/Engine/Ratings.cs
using VolleySim.Config;
using VolleySim.Domain;

namespace VolleySim.Engine
{
    /// <summary>10종 스탯 + 신장을 판정용 복합 레이팅(0~100 스케일)으로 합성.</summary>
    public static class Ratings
    {
        public static double Serve(Player p, SimConfig.RatingWeights w)
        {
            return w.ServeFromServe * p.Stats.Serve + w.ServeFromPower * p.Stats.Power;
        }

        public static double Receive(Player p, SimConfig.RatingWeights w)
        {
            return w.ReceiveFromReceive * p.Stats.Receive + w.ReceiveFromSpeed * p.Stats.Speed;
        }

        public static double Set(Player p, SimConfig.RatingWeights w)
        {
            return w.SetFromSet * p.Stats.Set + w.SetFromSpeed * p.Stats.Speed;
        }

        public static double Attack(Player p, SimConfig.RatingWeights w)
        {
            double h = SimMath.Clamp((p.HeightCm - w.HeightPivotCm) * w.AttackHeightPerCm, -w.AttackHeightCap, w.AttackHeightCap);
            return w.AttackFromSpike * p.Stats.Spike + w.AttackFromPower * p.Stats.Power + w.AttackFromSpeed * p.Stats.Speed + h;
        }

        public static double Block(Player p, SimConfig.RatingWeights w)
        {
            double h = SimMath.Clamp((p.HeightCm - w.HeightPivotCm) * w.BlockHeightPerCm, -w.BlockHeightCap, w.BlockHeightCap);
            return w.BlockFromBlock * p.Stats.Block + w.BlockFromSpeed * p.Stats.Speed + h;
        }

        public static double Dig(Player p, SimConfig.RatingWeights w)
        {
            return w.DigFromDig * p.Stats.Dig + w.DigFromSpeed * p.Stats.Speed;
        }
    }
}

File: web/engine/ratings.js
// 10종 스탯 + 신장 → 판정용 복합 레이팅(0~100). Engine/Ratings.cs 포팅.
import { STAT } from './domain.js';
import { clamp } from './mathx.js';

export function ratingServe(p, w) {
  return w.serveFromServe * p.stats[STAT.serve] + w.serveFromPower * p.stats[STAT.power];
}

export function ratingReceive(p, w) {
  return w.receiveFromReceive * p.stats[STAT.receive] + w.receiveFromSpeed * p.stats[STAT.speed];
}

export function ratingSet(p, w) {
  return w.setFromSet * p.stats[STAT.set] + w.setFromSpeed * p.stats[STAT.speed];
}

export function ratingAttack(p, w) {
  const h = clamp((p.heightCm - w.heightPivotCm) * w.attackHeightPerCm, -w.attackHeightCap, w.attackHeightCap);
  return w.attackFromSpike * p.stats[STAT.spike] + w.attackFromPower * p.stats[STAT.power]
       + w.attackFromSpeed * p.stats[STAT.speed] + h;
}

export function ratingBlock(p, w) {
  const h = clamp((p.heightCm - w.heightPivotCm) * w.blockHeightPerCm, -w.blockHeightCap, w.blockHeightCap);
  return w.blockFromBlock * p.stats[STAT.block] + w.blockFromSpeed * p.stats[STAT.speed] + h;
}

export function ratingDig(p, w) {
  return w.digFromDig * p.stats[STAT.dig] + w.digFromSpeed * p.stats[STAT.speed];
}
